service/history/shardController.go

```diff
@@ -167,18 +165,24 @@ func (c *shardController) getEngineForShard(shardID int) (Engine, error) {
 	if err != nil {
 		return nil, err
 	}
-	return item.getOrCreateEngine(c.shardClosedCh)
+	return item.getOrCreateEngine(c.shardClosedCallback)
 }
 
-func (c *shardController) removeEngineForShard(shardID int) {
+func (c *shardController) removeEngineForShard(shardID int, shardItem *historyShardsItem) {
 	sw := c.metricsScope.StartTimer(metrics.RemoveEngineForShardLatency)
 	defer sw.Stop()
 	item, _ := c.removeHistoryShardItem(shardID)
-	if item != nil {
+	if item != nil && (item == shardItem || shardItem == nil) {
```

Comment:
What is the case to compare the shardItem and item?

Reply:
It is just a defensive check to make sure we are deleting what we intend to delete. As for the specific scenario, consider the following:

(1) ShardContext gets an OwnershipLostError and decides to close shard - so it is about to call this method
(2) In the meanwhile, some other piece of code also calls removeEngineForShard - say, the frontend API for CloseShard()
(3) After (2) is done but before (1) is executed, acquireShards loads the same shard once again because now, we are supposed to own that shard (because of some other host going down)
(4) Now, (1) executes and without the defensive check, it will remove the wrong shardItem. Its not the end of the world, since we will recreate the shard on the next request or so  - but it doesn't hurt to have this additional check. I will add comment to clarify this.

Reply:
make sense to me now